security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisImpl.java

```diff
@@ -88,48 +89,76 @@ public boolean removeInterceptor(SecurityAnalysisInterceptor interceptor) {
-                                        network.getVariantManager().setWorkingVariant(postContStateId);
-
-                                        // apply the contingency on the network
-                                        contingency.toTask().modify(network, computationManager);
-
-                                        return null;
-                                    }, computationManager.getExecutor())
-                                    .thenComposeAsync(aVoid -> LoadFlow.runAsync(network, postContStateId, computationManager, postContParameters), computationManager.getExecutor())
-                                    .handleAsync((lfResult, throwable) -> {
-                                        network.getVariantManager().setWorkingVariant(postContStateId);
-
-                                        resultBuilder.contingency(contingency)
-                                                .setComputationOk(lfResult.isOk());
-                                        violationDetector.checkAll(contingency, network, resultBuilder::addViolation);
-                                        resultBuilder.endContingency();
-                                        network.getVariantManager().removeVariant(postContStateId);
-
-                                        return null;
-                                    }, computationManager.getExecutor());
+                        int workerCount = Math.min(computationManager.getResourcesStatus().getAvailableCores(), contingencies.size());
+                        BlockingQueue<Integer> queue = new ArrayBlockingQueue<>(workerCount);
+                        for (int i = 0; i < workerCount; i++) {
+                            String postContStateId = hash + "_" + i;
+                            network.getVariantManager().cloneVariant(VariantManagerConstants.INITIAL_VARIANT_ID, postContStateId);
+                            queue.add(i); //Will always work because we are initializing the queue and the initial capacity is workerCount
                         }
+                        // use completedFuture(null).thenCompose so that more of the
+                        // execution is handled by the whenComplete block removing the
+                        // variants.
+                        future = CompletableFuture.completedFuture(null).thenComposeAsync(aVoid -> {
+                            boolean previousMultiThreadAcces = network.getVariantManager().isVariantMultiThreadAccessAllowed();
+                            network.getVariantManager().allowVariantMultiThreadAccess(true);
+                            CompletableFuture<Void>[] futures = new CompletableFuture[contingencies.size()];
+                            for (int i = 0; i < contingencies.size(); i++) {
+                                Contingency contingency = contingencies.get(i);
+
+                                // run one loadflow per contingency
+                                futures[i] = CompletableFuture
```

Comment:
I think sonar cognitive complexity analyzer must be crying :)

It was already the case before the PR, but I think we should stick to the rule that lambda should be as much as possible one-liners, in particular for complex, asynchronous code. 
I think we should simply create more functions here.

Reply:
Yes, we should. I didn't refactor to allow to compare to the previous version. If we decide to go this way, we should refactor